Add calibration file input to ISIS Reflectometry GUI

rbauststfc opened the pull request:
**Description of work.**
This PR adds a new `CalibrationFile` input to the Instrument Settings tab of the Reflectometry interface. This input is blank by default, but when a file path is provided a calibration step will automatically be applied for normal reductions, live reductions or when loading preview data.

If the Debug checkbox on the Experiment Settings tab is ticked then the calibration step will output a table workspace containing the calibration data that is being applied. If the Debug checkbox is unticked then the calibration table workspace is removed from the ADS before the calibration step completes.

Although the calibration file path is optional data, I didn't use a `boost::optional` for holding the data in the Instrument model as it seemed sufficient to hold it as an empty string.

**Report to:** Max at ISIS

**To test:**

1) Open the ISIS Reflectometry interface and set the instrument to INTER.
2) On the Experiment Settings tab, tick the Debug checkbox. This will make sure that a calibration table is output to the ADS when the calibration step is run. Also set the AnalysisMode to `MultiDetectorAnalysis`.
3) Download the example calibration data from issue #32022. On the Instrument Settings tab, find the CalibrationFile input at the bottom of the tab and click the button to browse to the calibration file provided in the example data (`pixel_centers_repeat.dat`).
4) Go to the Runs tab. In the Runs table (on the right hand side) enter run `61135` and angle `0.5` into the first row of the table. Select the row and click to process. The reduction should complete successfully and the calibration step should have been applied (see messages in the panel, output calibration workspace and also `ReflectometryISISCalibration` should be in the history of workspace `IvsQ_binned_61135`).
5) Clear all the workspaces and go to the Reduction Preview tab. Load the `INTER45455_inst` dataset and again check that the calibration step is applied (there should be messages in the panel plus an output calibration workspace).
6) Finally, check live data reduction by clicking the Start Monitor button on the Runs tab (this will need to be tested at ISIS while instruments are live). The calibration step may throw errors if the test calibration file doesn't contain the right entries for the experiment, however that's sufficient to prove that calibration is being applied to live data reductions.
7) Test running reductions with the Debug checkbox unticked, which should prevent the calibration table workspace being left in the ADS. Also check that calibration is not applied when the CalibrationFile input is left blank.

Fixes #34111.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

rbauststfc committed:
Add GUI input for providing path to calibration file

rbauststfc committed:
Set up passing calibration file path between model and view

rbauststfc committed:
Pass calibration filepath to load and process algorithm from GUI

rbauststfc committed:
Pass calibration file from gui to preprocessing algorithm

rbauststfc committed:
Add calibration filepath validation to algorithm

rbauststfc committed:
Add file path validation to GUI input

rbauststfc committed:
Only output calibration table ws in debug mode

rbauststfc committed:
Edit list of related algorithms in calibration algorithm

rbauststfc committed:
Add release notes for issue 34111

rbauststfc committed:
Fix parameter being passed by value

rbauststfc committed:
Add calibration file path input to batch save and load

rbauststfc committed:
Remove explicit copy assignment operator

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/InstrumentPresenter.cpp:
```diff
@@ -42,6 +44,20 @@ void InstrumentPresenter::notifyRestoreDefaultsRequested() {
 }
 
+void InstrumentPresenter::notifyEditingCalibFilePathFinished() {
+  auto const calibrationFilePath = m_view->getCalibrationFilePath();
```
There's an interface `IReflMessageHandler` with a method `askUserForLoadFileName` you can use for getting the filename. (It's a work in progress but I've been trying to refactor out direct uses of QFileDialog.) It will require a bit of boilerplate to pass it in here but I think it's worth it. It can also be used for pop-up messages, so we can avoid direct use of QMessageBox too.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/InstrumentPresenter.cpp:
```diff
@@ -42,6 +44,20 @@ void InstrumentPresenter::notifyRestoreDefaultsRequested() {
+  auto const calibrationFilePath = m_view->getCalibrationFilePath();
+  if (!calibrationFilePath.empty()) {
+    try {
+      auto pocoPath = Poco::Path().parseDirectory(calibrationFilePath);
```
Similarly there's an `IFileHandler` interface - I don't think it currently has functionality to check if a file exists but I think that would be a good place to put these checks; then if they're needed in other parts of the GUI we can re-use them via the same interface, and easily mock them etc.

To my shame, the concrete type for both of these interfaces is the `QtMainWindow`! They should really be separated out into separate classes. But because we're using the interfaces, that should be a pretty easy maintenance job to do at some point.

rbauststfc replied:
Sounds good, thanks, I'll make that change. It looks like aren't unit tests for the FileHandler methods at the moment, I assume because they're defined in the view, or maybe because the nature of the existing methods would be hard to test? Assuming I haven't missed them somewhere, adding unit tests for the new method I'll add is something that I guess we could do as part of moving things into their own classes.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/QtInstrumentView.cpp:
```diff
@@ -93,6 +95,15 @@ void QtInstrumentView::disconnectSettingsChange(QCheckBox &edit) {
 void QtInstrumentView::onSettingsChanged() { m_notifyee->notifySettingsChanged(); }
 
+void QtInstrumentView::browseToCalibrationFile() {
+  auto calibrationFilePath = QFileDialog::getOpenFileName(this, QString(), QString(), tr("Data Files (*.dat)"));
```
As mentioned, I think we can use an interface to get the file path rather than direct use of QFileDialog, so here I suggest we just notify the presenter and let it do all the work, and then just call back to update the view with the result.

gemmaguest reviewed (changes requested):
This looks really good. Just a few code suggestions in-line, one thing that looks like it may be a bug, and one usability suggestion (possibly for a future PR).

**Possible bug**
If I run on a group workspace, e.g. POLREF37125, I get the following error. Looking at line 604 in ReflectometryISISLoadAndProcess this function is trying to get the run number for a workspace - I'm not sure why the workspace is none here! Might be worth checking this on main in case it's an unrelated problem.
```
Error in execution of algorithm ReflectometryISISLoadAndProcess:
'NoneType' object has no attribute 'getName'
  at line 84 in '/home/gemma/work/checkout/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 236 in '/home/gemma/work/checkout/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 424 in '/home/gemma/work/checkout/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 604 in '/home/gemma/work/checkout/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
```

**Usability suggestion**
When the user enters an invalid filename, it would be preferable to highlight the field in red rather than show a pop-up box. The pop-up keeps re-appearing whenever the focus changes between widgets, which can be a bit disruptive. This could be done in a future PR though if that is a big change. You can see an example of the red highlighting if you change a min to be greater than max e.g.
![image](https://user-images.githubusercontent.com/12895056/205131713-1d6de0c2-9ebf-49da-a97f-48cd0bf7eaa6.png)

**Testing comments**
Otherwise this functions really well and all interactions have been considered thoroughly. Great work :+1: :+1: :+1:
- Code changes look really good
- Works for all cases described (standard reduction, preview reduction and live data)
- Saving and loading batches works, and is backwards compatible.
- Good test coverage added; nice to see some tests for group workspaces too, which I always forget!

rbauststfc commented:
Good spot on the workspace group bug. This does seem to be to do with the calibration step, although was introduced by the previous PR that hooked up the calls between the various algorithms. It looks like the calibration algorithm itself produces an output for workspace groups (you can check this by calling `self.getProperty('OutputWorkspace').value` at the very end of the calibration algorithm, which returns a value). When `ReflectometryISISPreprocess` calls `alg.getProperty('OutputWorkspace').value` to get the output from the calibration algorithm, though, the returned value seems to be empty. It will need a bit of investigating, unless you've come across anything like this before?

rbauststfc committed:
Move filepath validation logic to IFileHandler

rbauststfc committed:
Use message handler for browsing to calibration file

rbauststfc committed:
Improve calibration file path validation

rbauststfc committed:
Update unit tests for validation changes

rbauststfc commented:
As an update, I've made some changes to address all the points made apart from fixing the bug. I'll now move on to look into that. Although this PR doesn't introduce the bug, it will make it a lot more discoverable so I would recommend we fix it before merging this in.

gemmaguest commented:
Playing around with running the algorithm as a child from a script (see below) I get slightly different behaviour with a workspace group - the result gets put in the ADS, whereas with a single workspace input it doesn't. I wonder if this is related to the problem we're seeing - perhaps when it puts it in the ADS it doesn't set it properly on the output property. If so this sounds like maybe a general bug with workspace groups!

I wonder if it would fix it if it didn't get put in the ADS. We can probably avoid this if we run all the child algorithms of ReflectometryISISCalibration as child algorithms too.

```
ws_in = Load("INTER63583")
alg = AlgorithmManager.create("ReflectometryISISCalibration")
alg.setChild(True)
alg.enableHistoryRecordingForChild(True)
alg.setProperties({"InputWorkspace":ws_in, "CalibrationFile":'/home/gemma/work/issues/refl_calibration/pixel_centers_repeat.dat', "OutputWorkspace":"ws_calibrated_child"})
alg.execute()


#
# Run as a child with ws group - puts the result in the ADS
#
ws_gp_in = Load("POLREF37125")
alg_gp = AlgorithmManager.create("ReflectometryISISCalibration")
alg_gp.setChild(True)
alg.enableHistoryRecordingForChild(True)
alg_gp.setProperties({"InputWorkspace":ws_gp_in, "CalibrationFile":'/home/gemma/work/issues/refl_calibration/pixel_centers_repeat_polref.dat', "OutputWorkspace":"ws_gp_calibrated_child"})
alg_gp.execute()
```

rbauststfc commented:
I think this may have something to do with passing in a workspace/workspace group as the input. I've tried the following:

1) I edited the calibration algorithm so that the only thing it did was take the workspace input and set it to the OutputWorkspace property. I still saw the behaviour that calling `alg.getProperty('OutputWorkspace').value` after passing in a workspace group returned None. If just a workspace was passed in then `alg.getProperty('OutputWorkspace').value` returned a value.

2) I changed the input property of the calibration algorithm to take a string value rather than a workspace and then edited the rest of the algorithm so that it simply loaded the workspace from the string value and set it to the OutputWorkspace. In this case calling `alg.getProperty('OutputWorkspace').value` returned a value for both single workspaces and workspace groups.

3) I repeated the same steps with the `ReflectometryISISPreprocess` algorithm, to check that there was nothing specific to the calibration algorithm setup that might be causing the behaviour, and I saw the same result.

rbauststfc committed:
Update ReflectometryISISLoadAndProcess documentation

rbauststfc committed:
Add link to algorithm documentation in release notes

gemmaguest commented:
> 4\. I repeated the second test but this time keeping the input workspace property as well as having a property that took a string. Internally the algorithm was simply loading the workspace from the string value and setting it to the OutputWorkspace (nothing was done with the input workspace). When the input workspace property (declared as either optional or mandatory) was set the results here were the same as for the first test, so it seemed just having a set input workspace property was affecting the behaviour, even though it's value was never used. When the input workspace property was optional but unset the results were the same as the second test.

Ok I think I've figured out what is happening here. If a workspace group is passed as an input then `Algorithm::processGroups` method is called because we know it is a group. If we just pass a string then it doesn't know it's a group; when we load it internally it loads as a group and `processGroups` will be called on each child method. 

This seems to work fine normally - ReflectometryISISLoadAndProcess and ReflectometryISISPreprocess both take string inputs and their child algorithms are called via `processGroups`. However, most of the child algorithms are C++; I wonder if there is a problem calling `processGroups` on python algorithms. (The only python child algorithms are ReflectometrySliceEventWorkspace and ReflectometryISISSumBanks, but neither is supported for workspace groups.)

gemmaguest commented:
I think for now we should just throw a "not supported" error for workspace groups if calibration is requested. Probably the best place to do this is in the _loadRun method in ReflectometryISISPreprocess

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rbauststfc committed:
Raise error if try to calibrate a workspace group via pre-processing

rbauststfc commented:
> I think for now we should just throw a "not supported" error for workspace groups if calibration is requested. Probably the best place to do this is in the _loadRun method in ReflectometryISISPreprocess

@gemmaguest I've added a commit for this. I added the validation to the `_applyCalibration` method, as I felt it might make sense to keep it with the rest of the code that handles when and how to apply calibration. See what you think, but I'm happy to move it to the `_loadRun` method if preferred.

gemmaguest reviewed (approved):
Thanks for the fix for group workspaces - trying to apply a calibration to these now issues an error message, which is fine for now because we don't need to support these in this initial version. I've re-checked with normal workspaces and the calibration is still applied as expected so this is good to go. It's really good to get this in for the release - great work :+1:
